### DISCUSSION TITLE
A question about redirect to https with  'lb + apisix'

### DISCUSSION THREAD

**Discussion by Lewisyixin** [apache/apisix#6928] [Q&A]
Is there any way to redirect from http  to https on apisix which behind a lb  meanwhile use http to proxy between lb and apisix?
We want to use apisix to replace nginx. Our nginx conf is about below..
Our ssl certs are on lb.  lb's 80 port proxy to nginx 80  port and 443 port proxy to nginx 443 port.
We just want to achieve the following conditions:

1. ssl cert just on lb and nginx 443 port is just a http port
2. redirection of https  is completed by nginx(because some domain don't want auto redirection)

`server{
        listen 80;
        server_name www.test.com;
        rewrite ^(.*)$ https://$host$1 permanent;
}

server {
        listen 443 ;
        server_name www.test.com;
}
`

I tried to complete this on apisix but I found that  a different  of apisix is that  mutiple http port can't handle different configs  like behand. So is there any to accomplish the goal？

**Comment by tzssangglass** [comment]
You can have APISIX listen to ports 80 and 443

Then create two routes, where the vars configuration is

```shell
curl -i http://127.0.0.1:9080/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "host": "foo.com",
    "vars": [
        [
            "server_port",
            "==",
            80
        ]
    ]
}'
```

```shell
curl -i http://127.0.0.1:9080/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "host": "foo.com",
    "vars": [
        [
            "server_port",
            "==",
            443
        ]
    ]
}'
```

The server_port variable is used to determine which listening port the request is coming from, then configure redirect or other plugins on both routes to perform different actions.

**Reply by Lewisyixin** [reply]
before this, it seems auto redirection didn't work after proxy.
Our real data flow is two way: 1. aliyun slb 80 --> nginx 80 --> apisix 80; 2. aliyun slb 443(with only ssl cert on ) --> nginx 443(http) --> apisix 443(http)
In my opinoin, before I create two routes with different "server_port", this will fall into a redirection loop if I set "http_to_https" to true. But the actual situation seems that "http_to_https" didn't work. Here's the log:
1. I curl from nginx server to apisix, results are as expected that we got a 301 redirection
<img width="993" alt="image" src="https://user-images.githubusercontent.com/17898688/165347902-83556edf-ae4f-4d67-a737-b2b9630ccf44.png">
2. I curl from client to slb, I just got a 200 code
<img width="847" alt="image" src="https://user-images.githubusercontent.com/17898688/165348229-73984d87-adee-49b6-a708-e4788bd7832f.png">
the nginx log show that it proxy a http request to apisix:
<img width="1098" alt="image" src="https://user-images.githubusercontent.com/17898688/165348476-3264415c-cab4-4a9a-9573-43dfad5b3518.png">
and apisix log show that it got a http request
<img width="1100" alt="image" src="https://user-images.githubusercontent.com/17898688/165348635-692c24ae-d093-4705-a1d9-708d5d4be879.png">
So why request from nginx to apisix didn't get 301?
apisix route config is :
<img width="843" alt="image" src="https://user-images.githubusercontent.com/17898688/165349473-964277c8-6e7e-47a3-ae5a-2a566409e0b3.png">

**Comment by tzssangglass** [comment]
> I curl from nginx server to apisix, results are as expected that we got a 301 redirection
> So why request from nginx to apisix didn't get 301? apisix route config is :

I can't solve your problem, it looks like you are getting the correct response from nginx to APISIX, then the problem may be in the SLB.

**Reply by Lewisyixin** [reply]
I'm afraid not. When I curl slb. Look at the 4th picture, that's log on apisix of request from nginx.I think it should be a 301, not 200.

**Reply by Lewisyixin** [reply]
In other words, we ignore the slb or even nginx,  is it normal that apisix can get a 200 http response when the request route has already be configed "http_to_https" to true?
<img width="1431" alt="image" src="https://user-images.githubusercontent.com/17898688/165431418-32419974-3f2e-4bc7-b5be-fa50029bb458.png">

**Reply by Lewisyixin** [reply]
Oh, sry, I just found the reason by tcpdump. I use 'proxy_set_header X-Forwarded-Proto https;' on nginx. When I delete it, redirection  works well

**Reply by Lewisyixin** [reply]
@tzssangglass  Another question is that I found  setting "http_to_https" must have ssl certs first, or will get error:
<img width="377" alt="image" src="https://user-images.githubusercontent.com/17898688/165701539-708856a9-4df8-4b31-abe1-db264d5c707f.png">
In my scene "apisix after slb and redirection is on apsix" ,  I just don't wan't save ssl certs on apisix. So could SSL cert not be the necessary conditions for http_to_https?

**Reply by tzssangglass** [reply]
> In my scene "apisix after slb and redirection is on apsix" , I just don't wan't save ssl certs on apisix. So could SSL cert not be the necessary conditions for http_to_https?

 SSL cert not be the necessary conditions for http_to_https

I'm not sure that `SSL cert not exists for sni` is return by APISIX, I can't find any message like this in APISIX.

**Reply by tzssangglass** [reply]
You can give the complete reproduction steps and find where caused the problem is.

**Reply by Lewisyixin** [reply]
I got this error by using dashboard.  Just set a domain and  turn on "http_to_https" when creating a route rule will easily reproduce.
I also found this error is returned by dashbord.
https://github.com/apache/apisix-dashboard/blob/88f323285f2bdbafca027da9aef357b2e3eb4008/api/internal/handler/ssl/ssl.go
So should I feedback this in dashboard project?

**Reply by tokers** [reply]
@Lewisyixin It looks like the problem is caused by the dashboard. I can't transfer the discussion there, @tzssangglass Could you transfer it?
